frontend/src/scenes/surveys/surveyLogic.tsx

```diff
@@ -1038,6 +1039,58 @@ export const surveyLogic = kea<surveyLogicType>([
                 return SurveyQuestionBranchingType.ConfirmationMessage
             },
         ],
+        hasCycle: [
+            (s) => [s.survey],
+            (survey) => {
+                const graph = new Map()
+                survey.questions.forEach((question, fromIndex: number) => {
+                    if (!graph.has(fromIndex)) {
+                        graph.set(fromIndex, new Set())
+                    }
+
+                    if (
+                        question.branching?.type === SurveyQuestionBranchingType.SpecificQuestion &&
+                        Number.isInteger(question.branching.index)
+                    ) {
+                        const toIndex = question.branching.index
+                        graph.get(fromIndex).add(toIndex)
+                    } else if (
+                        question.branching?.type === SurveyQuestionBranchingType.ResponseBased &&
+                        isObject(question.branching?.responseValues)
+                    ) {
+                        for (const [_, toIndex] of Object.entries(question.branching?.responseValues)) {
+                            if (Number.isInteger(toIndex)) {
+                                graph.get(fromIndex).add(toIndex)
+                            }
+                        }
+                    }
+
+                    // No branching - still need to connect the next question
+                    if (fromIndex < survey.questions.length - 1) {
+                        const toIndex = fromIndex + 1
+                        graph.get(fromIndex).add(toIndex)
+                    }
+                })
+
+                let cycleDetected = false
+                function dfs(node: number, seen: number[]): void {
+                    if (cycleDetected) {
+                        return
+                    }
+
+                    for (const neighbor of graph.get(node) || []) {
+                        if (seen.includes(neighbor)) {
```

Comment:
this isn't sufficient to detect cycles, you need a recursionStack that detects not just visited but in the current stack for the start of a cycle. For example, this case:

Q1 has branching to Q2 and Q3.

Q2 branches to Q4 and **Q3**

Q3 branches to Q5.

In this case, there's no cycles, Q1->2->3->5 is valid, so is Q1->3->5 and Q1->2->4 but since Q3 is visited 2x, this detects a cycle when it isn't.

More details (didn't actually read this but hoping its roughly correct 🙈  -> https://chatgpt.com/share/f3eac772-e2fd-424a-9cef-1e5255e20a1f )

Reply:
But this does work correctly 🙈 - test added.

![image](https://github.com/PostHog/posthog/assets/22996112/ed8b7a5f-bf5b-4ec4-8532-46fa7e0bb79b)

This works because every recursion call receives its own copy of `seen` (because using concat). So all paths are traveled, and every path has its own array of nodes that have been seen. The cycle is found if the same node appears twice.

Reply:
I did however find other issues:
- When building the graph, if the branching type is `confirmation_message` or `specific_question`, we should not connect the current node to Next question - fixed
- If response-based branching is exhaustive (i.e. explicitly defining all possible options/buckets), we shouldn't connect the current node to the next question. I considered fixing this but couldn't find an example where it caused incorrect cycle detection/non-detection. Fixing this overcomplicates the code and requires updating if we introduce more scales/question types, so for now I would handwave this, especially because this graph checking is not any kind of safety-critical thing. Curious if we ever encounter issues in production.

Fun fact: Hotjar lets you save surveys with cycles 😅

Reply:
lol, very nice, indeed I missed that in my mental simulation of the dfs algo - nice! (smh chatgpt)

I'm wondering if the implicit to next question branching when there are other branching types defined is prone to bugs. Should this be explicit? But then again, if we match nothing we go to the next question anyway so 🤷 , guess we'll find out, don't have a strong opinion here